archimedes/archimedes/runner.py

```diff
@@ -0,0 +1,93 @@
+import os.path
+import os
+import threading
+from dataclasses import dataclass
+from hashlib import md5
+from typing import List, Iterable
+from docker import DockerClient
+from docker.types import Mount
+from archimedes.checker import run as run_checker
+import tempfile
+
+
+@dataclass
+class RunResult:
+    status: str
+    error: List[str]
+
+class Runner:
+    def open_key(self, key, mode):
+        # TODO, this is just a demonstration.  Ideally this would be a input/output stream to
+        # external storage service.
+        if mode.startswith("r"):
+            return open(os.path.join(os.path.dirname(__file__), "..", "tests", "fixtures", "runner_inputs", key), mode)
+        return tempfile.TemporaryFile(mode)
+
+    # A simple entrypoint demonstrating a process to run.
+    def run(self, script_text, image="vulcan_base", output_keys: Iterable[bin] = tuple(), input_keys: Iterable[bin] = tuple(), environment: Iterable[str] = tuple()):
+        """
+        :param script_text: The python script's text contents to run.
+        :param output_keys: Keys to store process outputs. Any iterable.
+        :param input_keys:  Keys to find process inputs. Any iterable.
+        :param environment: Iterable of strings in the form "MYENVKEY=myvalue"
+        :return: For now, nothing, but in the future, this function may become asynchronous and
+            attach to a scheduler, in which case it would return the status of the job being scheduled.
+        """
+        with tempfile.TemporaryDirectory() as inputs_dir, tempfile.TemporaryDirectory() as outputs_dir, tempfile.TemporaryDirectory() as exec_dir:
+            script_path = os.path.join(exec_dir, "script.py")
+            with open(script_path, "w") as script_file:
+                script_file.write(script_text)
+
+            if not run_checker([script_path]):
+                raise ValueError("Script did not conform to the archimedes dsl requirements.")
+
+            mounts = [
+                Mount(
+                    target="/var/run/vulcan/exec",
+                    source=exec_dir,
+                    read_only=True
+                ),
+                Mount(
+                    target="/var/run/vulcan/inputs",
+                    source=inputs_dir,
+                    read_only=True
+                ),
+                Mount(
+                    target="/var/run/vulcan/outputs",
+                    source=outputs_dir,
+                    read_only=False
+                ),
+            ]
+
+            output_pipes = [
+                (os.mkfifo(os.path.join(outputs_dir, md5(output_key).hexdigest())), output_key)
```

Comment:
This ensures that any inputs and outputs are streamed and do not buffer -- even if the storage is remote, like redis, the inputs and outputs would only occupy small chunks at a time as they make their way over the network.